scripts/datastore_upgrade.py

```diff
@@ -0,0 +1,330 @@
+
+  except Exception as ex:
+    logging.info("Exception on range query: {}".format(str(ex)))
+    sys.exit(1)
+
+  # Remove the tombstoned entities from the list.
+  tombstoneless_entities = ds_distributed.remove_tombstoned_entities(current_entities)
+  app_ids = get_app_ids_from_entities(tombstoneless_entities)
+  entities_to_validate  = tombstoneless_entities
+
+  # Validate the entities grouped by app id.
+  for app_id in app_ids:
+    validated_entities_for_app = ds_distributed.validated_list_result(app_id, entities_to_validate)
+    entities_to_validate = validated_entities_for_app
+
+  return current_entities, tombstoneless_entities, validated_entities_for_app
+
+def delete_tombstoned_entities_from_table(datastore, current_entities, validated_entities):
+  """ Deletes the set of rows corresponding to entity keys which have
+  tombstoned values from the APP_ENTITY_TABLE.
+  Args:
+    datastore: A reference to the batch datastore interface.
+    current_entities: A list of entities currently in the APP_ENTITY_TABLE.
+    validated_entities: A list of validated entities to compare from.
+  """
+  valid_entity_keys = get_valid_entity_keys(validated_entities)
+  invalid_entity_keys = compare_and_get_invalid_keys(current_entities, valid_entity_keys)
+  try:
+    datastore.batch_delete(APP_ENTITY_TABLE, invalid_entity_keys)
+    logging.info("Batch delete was successful")
+  except AppScaleDBConnectionError as error:
+    logging.info("Exception on batch delete: {}".format(str(error)))
+    sys.exit(1)
+
+def update_entities_table(datastore, entities, validated_entities):
+  """ Updates the APP_ENTITY_TABLE with valid entities if any.
+  Args:
+    datastore: A reference to the batch datastore interface.
+    entities: A list of tombstonedless entities.
+    validated_entities: A list of validated entities to compare from.
+  """
+  row_keys_to_update = []
+  cell_values_to_update = {}
+  for valid_entity in validated_entities:
+    valid_entity_key = valid_entity.keys()[0]
+    for entity in entities:
+      entity_key = entity.keys()[0]
+      if not valid_entity_key == entity_key:
+        continue
+
+      if is_entity_same(entity,entity_key, valid_entity) and is_txn_id_same(entity, entity_key, valid_entity):
+        continue
+
+      if not entity_key in row_keys_to_update:
+        row_keys_to_update.append(entity_key)
+        cell_value = {}
+        cell_value[APP_ENTITY_SCHEMA[0]] = valid_entity[entity_key][APP_ENTITY_SCHEMA[0]]
+        cell_value[APP_ENTITY_SCHEMA[1]] = valid_entity[entity_key][APP_ENTITY_SCHEMA[1]]
+        cell_values_to_update[entity_key] = cell_value
+
+  try:
+    datastore.batch_put_entity(APP_ENTITY_TABLE, row_keys_to_update, APP_ENTITY_SCHEMA, cell_values_to_update)
+    logging.info("Batch put entities was successful")
+  except  AppScaleDBConnectionError as error:
+    logging.info("Exception on batch put entity: {}".format(str(error)))
```

Comment:
Make a pass on these and change the log level (info, debug, warning, error, exception) when appropriate.